README
==Redirect

A Ruby redirect server build on Rack

==Example

% ruby example.rb

History.txt
=== 0.3.2 / 2009-03-03

* 1 major enhancement

  * Added generation of tests to the app generator

=== 0.3.0 / 2009-02-24

* 1 major enhancement

  * Added reuse of match patterns
    [/old\/(.*)/, '/new/$1']  # /old/2008/02/01 will be redirected to /new/2008/02/01

=== 0.2.0 / 2009-02-22

* 1 major enhancement

  * Added a generator method
  
=== 0.1.0 / 2009-02-22

* 1 major enhancement

  * Works properly with rackup now
  
=== 0.0.4 / 2009-02-21

* 1 minor bug fix

  * Removed Hoe dependency
  
=== 0.0.3 / 2009-02-21

* 1 minor bug fix

  * Fixed gemspec
  
=== 0.0.2 / 2009-02-21

* 1 major enhancement

  * Only require Hoe if installed
  
=== 0.0.1 / 2009-02-08

* 1 major enhancement

  * Birthday!


spec/helper.rb
require 'rubygems'
require 'test/unit'
require 'test/spec'
require 'mocha'

$:.unshift File.dirname(File.dirname(__FILE__)) + "/lib"
require 'redirect'

spec/rack_spec.rb
require File.dirname(__FILE__) + '/helper'

describe "Rack::Redirect" do

  it "should show 404 not found with if no redirect exists" do
    @app = Rack::Redirect.new()
    res = Rack::MockRequest.new(@app).get("/test")

    res.not_found?.should be_true
    res["Content-Type"].should == "text/html"
    res.body.should == "not found"
  end
  
  it "should show index if no redirect exists but index does" do
    @app = Rack::Redirect.new()
    @app.index = "index"
    res = Rack::MockRequest.new(@app).get("/")
    # res.not_found?.should be_true
    res["Content-Type"].should == "text/html"
    res.body.should == "index"
  end  
  
  it "should redirect '/' if redirect exists" do
    @app = Rack::Redirect.new(['/', '/test'])
    res = Rack::MockRequest.new(@app).get('/')
    res.headers.should == { 'Location' => '/test', 'Content-Type' => 'text/html' }
    res.body.should == 'Redirecting to: /test'
  end  
  
  it "should show sitemap.xml" do
    @app = Rack::Redirect.new(['/', '/test', {:name => 'test'}])
    res = Rack::MockRequest.new(@app).get('/sitemap.xml')
    res.headers.should == {'Content-Type' => 'text/xml' }
    res.body.should == %(<?xml version="1.0" encoding="UTF-8"?>\n) +
    %(<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n) +
    %(<url>\n) +
      %(<loc>http://example.org/test</loc>\n) +
      %(</url>\n) +
    %(</urlset>\n)
  end
  
  it "should show sitemap.xml only for named redirects" do
    @app = Rack::Redirect.new(['/', '/test', {:name => 'test'}], ['/a', '/hidden'])
    res = Rack::MockRequest.new(@app).get('/sitemap.xml')
    res.body.should == %(<?xml version="1.0" encoding="UTF-8"?>\n) +
    %(<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n) +
    %(<url>\n) +
      %(<loc>http://example.org/test</loc>\n) +
      %(</url>\n) +
    %(</urlset>\n)
  end  
  
  it "should redirect if redirect exists" do
    @app = Rack::Redirect.new(['/old/2008', '/2008-new'])
    res = Rack::MockRequest.new(@app).get('/old/2008')
    res.headers.should == { 'Location' => '/2008-new', 'Content-Type' => 'text/html' }
    res.body.should == 'Redirecting to: /2008-new'
  end

  it "should redirect for a regular_expression" do
    @app = Rack::Redirect.new(['^/old', '/new'])
    res = Rack::MockRequest.new(@app).get('/old/2008')
    res.headers.should == { 'Location' => '/new', 'Content-Type' => 'text/html' }
    res.body.should == 'Redirecting to: /new'
  end

  it "should redirect for a regular_expression with rewrite" do
    @app = Rack::Redirect.new([/old\/(.*)/, '/new/$1'])
    res = Rack::MockRequest.new(@app).get('/old/2008/02/14')
    res.headers.should == { 'Location' => '/new/2008/02/14', 'Content-Type' => 'text/html' }
    res.body.should == 'Redirecting to: /new/2008/02/14'
  end

  it "should allow multiple redirects" do
    @app = Rack::Redirect.new(['^/old3', '/new'], ['^/old2', '/new2'])
    res = Rack::MockRequest.new(@app).get('/old2008')
    res.headers.should == { 'Location' => '/new2', 'Content-Type' => 'text/html' }
    res.body.should == 'Redirecting to: /new2'
  end

  it "should turn redirects array into Redirect Objects" do
    @app = Rack::Redirect.new(['^/old3', '/new', {:code => 307}], ['^/old2', '/new2', {:name => 'test'}])
    res = Rack::MockRequest.new(@app).get("/test")

    res.not_found?.should be_true
    res["Content-Type"].should == "text/html"
    res.body.should == "not found"
  end

end

describe "Redirect::Data" do
  it "should turn an array into an Object" do
    data = Redirect::Data.new('^/old3', '/new', {:code => 307, :name => 'test'})
    data.catch_url.should == '^/old3'
    data.redirect_url.should == '/new'
    data.code.should == 307    
    data.name.should == 'test'
  end
  
  it "should turn an dafault array into an Object" do
    data = Redirect::Data.new('^/old3', '/new')
    data.catch_url.should == '^/old3'
    data.redirect_url.should == '/new'
    data.code.should == 301
    data.name.should == nil
  end  
end

describe "Redirect" do
  after do
    Redirect.default_code = 301
    Redirect.app = nil
    Redirect.autorun = true
  end
  
  it "Should be able to configure the default http code" do
    data = Redirect::Data.new('/a', '/b')
    data.code.should == 301    
    Redirect.default_code = 307
    data = Redirect::Data.new('/a', '/b')
    data.code.should == 307
  end
  
  it "Should be able to set app" do
    Redirect.app.should == nil
    Redirect.app = 'a'
    Redirect.app.should == 'a'
  end  

  it "Should be able to set autorun" do
    Redirect.autorun.should == true
    Redirect.autorun = false
    Redirect.autorun.should == false
  end  
  
end

lib/redirect/test.rb
require 'redirect'
require 'test/unit'

class Test::Unit::TestCase

  def assert_redirects(from, to)
    res = Rack::MockRequest.new(APP[:redirect_app]).get(from)
    assert_equal(res.headers, { 'Location' => to, 'Content-Type' => 'text/html' })
  end

end

APP = {}

def redirect(*redirects)
  APP[:redirect_app] = Rack::Redirect.new(*redirects)
end

lib/redirect.rb
require 'rubygems'
require 'rack'
require 'rack/showexceptions'
require 'rack/request'
require 'rack/response'

module Redirect
  VERSION = '0.3.2'
  
  def self.default_code= default_code
    @default_code = default_code
  end
  
  def self.default_code
    @default_code ||= 301
  end

  def self.autorun= autorun
    @autorun = autorun
  end
  
  def self.autorun
    @autorun = true unless @autorun == false
    @autorun
  end

  def self.app= app
    @app = app
  end
  
  def self.app
    @app
  end
  
  class Data
    attr_reader :catch_url, :redirect_url, :code, :name
    def initialize(catch_url, redirect_url, options = {})
      @catch_url = catch_url
      @redirect_url = redirect_url
      @code = options[:code] || Redirect.default_code
      @name = options[:name]
    end
  end
    
end

def redirect(*redirects)

  Redirect.app = Rack::Redirect.new(*redirects)
  if Redirect.autorun
    Rack::Handler::WEBrick.run \
      Rack::ShowExceptions.new(Rack::Lint.new(Redirect.app)),
      :Port => 3000
    run Redirect.app
  end
end

module Rack

  class Redirect
    attr_reader :redirects
    def initialize(*redirects)
      @redirects = redirects.collect do |r|
        ::Redirect::Data.new(*r)
      end
    end
    
    def call(env)
      req = Request.new(env)
      if req.fullpath == '/sitemap.xml'
        return [200, {"Content-Type" => "text/xml"}, sitemap(req.host)]
      end
      if req.fullpath == '/' && index
        return [200, {"Content-Type" => "text/html"}, index]
      end      
      @redirects.each do |r|
        if req.fullpath.match(r.catch_url)
          redirect_url = r.redirect_url
          if $1
            redirect_url.gsub!('$1', $1)
          end
          puts "Match found for #{r.catch_url}."
          puts "Redirecting to #{redirect_url}"
          return [r.code, {"Location" => redirect_url, "Content-Type" => "text/html"}, "Redirecting to: #{redirect_url}"]
        end
      end
      [404, {"Content-Type" => "text/html"}, "not found"]
    end
   
    def sitemap(host)
      %(<?xml version="1.0" encoding="UTF-8"?>\n) +
      %(<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n) +
      @redirects.select{|r| r.name }.collect { |r|
        %(<url>\n) +
          %(<loc>http://#{host}#{r.redirect_url}</loc>\n) +
        %(</url>\n)}.join +
      %(</urlset>\n)
    end
    
    def index
      @index
    end
    
    def index= index
      @index = index
    end
    
  end
  
end